applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+// trailing_edge_sub_model_part
+void Define3DWakeProcess::AddTrailingEdgeAndWakeElements(std::vector<std::size_t>& rWakeElementsOrderedIds,
+                                                         std::vector<std::size_t>& rTrailingEdgeElementsOrderedIds) const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+
+    std::sort(rWakeElementsOrderedIds.begin(),
+              rWakeElementsOrderedIds.end());
+    root_model_part.GetSubModelPart("wake_elements_model_part").AddElements(rWakeElementsOrderedIds);
+
+    std::sort(rTrailingEdgeElementsOrderedIds.begin(),
+              rTrailingEdgeElementsOrderedIds.end());
+    ModelPart& trailing_edge_sub_model_part =
+        root_model_part.GetSubModelPart("trailing_edge_elements_model_part");
+    trailing_edge_sub_model_part.AddElements(rTrailingEdgeElementsOrderedIds);
+
+    std::vector<std::size_t> trailing_edge_nodes_ordered_ids;
+    // Add also the nodes of the elements touching the trailing edge
+    for (auto& r_elem : trailing_edge_sub_model_part.Elements())
+    {
+        for (auto& r_node : r_elem.GetGeometry()){
+            trailing_edge_nodes_ordered_ids.push_back(r_node.Id());
+        }
+    }
+
+    std::sort(trailing_edge_nodes_ordered_ids.begin(),
+              trailing_edge_nodes_ordered_ids.end());
+    trailing_edge_sub_model_part.AddNodes(trailing_edge_nodes_ordered_ids);
+}
+
+// This function recomputes the wake distances from the nodes belonging to te
+// elements. These distances are used later to decide which elements are KUTTA,
+// which are WAKE (STRUCTURE), and which are NORMAL
+void Define3DWakeProcess::RecomputeNodalDistancesToWakeOrWingLowerSurface() const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    ModelPart& trailing_edge_sub_model_part =
+        root_model_part.GetSubModelPart("trailing_edge_elements_model_part");
+
+    block_for_each(trailing_edge_sub_model_part.Nodes(), [&](Node<3>& r_node)
+    {
+        // Trailing edge nodes are assigned a positive distance
+        if(r_node.GetValue(TRAILING_EDGE)){
+            r_node.SetValue(WAKE_DISTANCE, mTolerance);
+        }
+        // Nodes in the lower surface are assigned a negative distance
+        else if(r_node.GetValue(LOWER_SURFACE)){
+            r_node.SetValue(WAKE_DISTANCE, -mTolerance);
+        }
+        // Nodes in the upper surface are assigned a positive distance
+        else if(r_node.GetValue(UPPER_SURFACE)){
+            r_node.SetValue(WAKE_DISTANCE, mTolerance);
+        }
+        // For the rest of the nodes the distance is recomputed:
+        else{
+            // TODO: improve the distance calculation.
+            // Find closest trailing edge node
+            NodeType::Pointer p_closest_te_node = *mrTrailingEdgeModelPart.NodesBegin().base();
+            FindClosestTrailingEdgeNode(p_closest_te_node, r_node);
+            RecomputeDistance(p_closest_te_node, r_node);
+        }
```

Comment:
Why is this needed exactly? Is it because there are some nodes that belong to elements touching the TE but that were not intersected (and thus do not have ELEMENTAL_DISTANCES)

Reply:
Yes exactly, that is one reason. Another reason is to make sure that the distance is unique in each point.

Reply:
@marcnunezc I forgot to ask you before whether you think that this is not necessary, we can discuss it next week. 

Still, I am aware of some corner cases in which the function RecomputeDistance would fail. I have an idea on how to improve the current implementation, but I would be interested in knowing whether you have other ideas on how to do this in general.